x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -127,12 +224,141 @@ func (k msgServer) queryAndValidateClaimForProof(ctx context.Context, proof *typ
 
 	// Ensure service IDs match.
 	if claimSessionHeader.GetService().GetId() != proofSessionHeader.GetService().GetId() {
-		return types.ErrProofInvalidService.Wrapf(
+		return nil, types.ErrProofInvalidService.Wrapf(
 			"claim service ID %q does not match proof service ID %q",
 			claimSessionHeader.GetService().GetId(),
 			proofSessionHeader.GetService().GetId(),
 		)
 	}
 
+	return &foundClaim, nil
+}
+
+// compareSessionHeaders compares a session header against an expected session header.
+func compareSessionHeaders(
+	expectedSessionHeader *sessiontypes.SessionHeader,
+	sessionHeader *sessiontypes.SessionHeader,
+) error {
+	if sessionHeader.GetApplicationAddress() != expectedSessionHeader.GetApplicationAddress() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders application addresses mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetApplicationAddress(),
+			sessionHeader.GetApplicationAddress(),
+		)
+	}
+
+	if sessionHeader.GetService().GetId() != expectedSessionHeader.GetService().GetId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders service IDs mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetService().GetId(),
+			sessionHeader.GetService().GetId(),
+		)
+	}
+
+	if sessionHeader.GetSessionStartBlockHeight() != expectedSessionHeader.GetSessionStartBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session start heights mismatch expect: %d, got: %d",
+			expectedSessionHeader.GetSessionStartBlockHeight(),
+			sessionHeader.GetSessionStartBlockHeight(),
+		)
+	}
+
+	if sessionHeader.GetSessionEndBlockHeight() != expectedSessionHeader.GetSessionEndBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session end heights mismatch expect: %d, got: %d",
+			expectedSessionHeader.GetSessionEndBlockHeight(),
+			sessionHeader.GetSessionEndBlockHeight(),
+		)
+	}
+
+	if sessionHeader.GetSessionId() != expectedSessionHeader.GetSessionId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session IDs mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetSessionId(),
+			sessionHeader.GetSessionId(),
+		)
+	}
+
+	return nil
+}
+
+// verifyClosestProof verifies the closest merkle proof against the expected root hash.
+func verifyClosestProof(
+	proof *smt.SparseMerkleClosestProof,
+	expectedRootHash []byte,
+) error {
+	spec := smt.NoPrehashSpec(sha256.New(), true)
```

Comment:
Can you move this into a local variable that's initiated on `init` and add a comment explaining it?

Seems like really important business logic that's embedded deep in the code.